pslab/external/hcsr04.py

```diff
@@ -0,0 +1,118 @@
+"""Ultrasonic distance sensors."""
+import time
+
+from pslab.instrument.logic_analyzer import LogicAnalyzer
+from pslab.instrument.waveform_generator import PWMGenerator
+from pslab.serial_handler import SerialHandler
+
+
+class HCSR04(LogicAnalyzer, PWMGenerator):
+    """Read data from ultrasonic distance sensor HC-SR04/HC-SR05.
+
+    These sensors can measure distances between 2 cm to 4 m (SR04) / 4.5 m
+    (SR05).
+
+    Sensors must have separate trigger and echo pins. First a 10 µs pulse is
+    output on the trigger pin. The trigger pin must be connected to the TRIG
+    pin on the sensor prior to use.
+
+    Upon receiving this pulse, the sensor emits a sequence of sound pulses, and
+    the logic level of its echo pin is also set high.  The logic level goes LOW
+    when the sound packet returns to the sensor, or when a timeout occurs.
+    Timeout occurs if no echo is received within the time slot determinded by
+    the sensor's maximum range.
+
+    The ultrasound sensor outputs a series of eight sound pulses at 40 kHz,
+    which corresponds to a time period of 25 µs per pulse. These pulses reflect
+    off of the nearest object in front of the sensor, and return to it. The
+    time between sending and receiving of the pulse packet is used to estimate
+    the distance. If the reflecting object is either too far away or absorbs
+    sound, less than eight pulses may be received, and this can cause a
+    measurement error of 25 µs which corresponds to 8 mm.
+
+    Parameters
+    ----------
+    device : :class:`SerialHandler`
+        Serial connection to PSLab device.
+    trig : str, optional
+        Name of the trigger pin. Defaults to SQ1.
+    echo : str, optional
+        Name of the echo pin. Defaults to LA1.
+
+    Example
+    -------
+    In this example the sensor's Vcc pin is connected to the PSLab's PV1 pin,
+    the Trig pin is connected to SQ1, Echo to LA1, and Gnd to GND.
+
+    >>> import pslab
+    >>> from pslab.external.hcsr04 import HCSR04
+    >>> psl = pslab.ScienceLab()
+    >>> distance_sensor = HCSR04(psl)
+    >>> psl.power_supply.pv1 = 5  # Power the sensor.
```

Comment:
One can also use `VCC` pin as the 5V source.

Reply:
Yes, or another SQ* pin set high. I prefer to leave the details of how to power the sensor to the user.